lib/tasks/bulk.rake

```diff
@@ -11,4 +11,80 @@ namespace :bulk do
       DeliveryRequestWorker.perform_async_in_queue(id, queue: :delivery_immediate)
     end
   end
+
+  desc "Switch immediate subscribers of the specified list slugs to daily digest"
+  task switch_to_daily_digest: :environment do |_t, args|
+    lists = SubscriberList.where(slug: args.extras)
+    raise "One or more lists were not found" if lists.count != args.extras.count
+
+    subscriptions = Subscription.active.immediately.where(subscriber_list: lists)
+    raise "No subscriptions to change" if subscriptions.none?
+
+    subscribers = Subscriber.where(id: subscriptions.pluck(:subscriber_id)).index_by(&:id)
+    subscriptions_by_subscriber = subscriptions.group_by(&:subscriber_id).transform_keys { |k| subscribers[k] }
+
+    subscriptions_by_subscriber.each do |subscriber, immediate_subscriptions|
+      subscriber.with_lock do
+        immediate_subscriptions.each do |subscription|
+          subscription.end(reason: :bulk_change)
+
+          Subscription.create!(
+            subscriber_id: subscription.subscriber_id,
+            subscriber_list_id: subscription.subscriber_list_id,
+            frequency: :daily,
+            source: :bulk_change,
+          )
+        end
+      end
+
+      email_id = SwitchToDailyDigestEmailBuilder.call(subscriber, immediate_subscriptions)
```

Comment:
Sure, I've done that. I think originally I was worrying about an actually-non-existent scenario where the email did send successfully, but some error occurred after (the point being the email isn't critical).

Reply:
Putting it inside the lock has the advantage that we're even less likely to send a confusing email in the presence of a race condition.

Reply:
Update: turns out the async sending of the email can be in a race with the creation of it inside the transaction. I've moved the last line out of the transaction/lock, to avoid this.